src/main/engine/Misc/src/GameInstance.cpp

```diff
@@ -245,24 +191,22 @@ int GameInstance::updateCameras() {
  -1 is returned and nothing is rendered.
 */
 int GameInstance::updateObjects() {
-    if (gameObjects.empty()) {
+    std::unique_lock<std::mutex> lock(sceneLock_);
+    if (sceneObjects_.empty()) {
         cerr << "Error: No active GameObjects in the current scene!\n";
         return -1;
     }
-    /// @todo: Polymorphism! Only need one structure for these game objects
-    glBindVertexArray(vertexArrayID);
-    std::vector<GameObject *>::iterator it;
-    for (it = gameObjects.begin(); it != gameObjects.end(); ++it) {
-        (*it)->setDirectionalLight(directionalLight);
-        (*it)->setLuminance(luminance);
-        // Send the new VP matrix to the current gameObject being drawn.
-        (*it)->setVpMatrix(getCamera((*it)->getCameraId())->getVpMatrix());
-        (*it)->render();
+    gfxController_->update();
+    // Update cameras first
+    vector<SceneObject *> deferredUpdates;
+    for (auto it = sceneObjects_.begin(); it != sceneObjects_.end(); ++it) {
+        if ((*it)->type() == ObjectType::CAMERA_OBJECT) (*it)->update();
+        else
```

Comment:
Polymorphism! We need to update cameras first to ensure rendering is not disturbed. If we update cameras while rendering other objects, we will see a flickering behavior. This isn't a race condition. It's due to rendering GameObjects in-between camera updates.